modules/core/shared/src/test/scala/retry/PackageObjectSpec.scala

```diff
@@ -216,6 +216,192 @@ class PackageObjectSpec extends AnyFlatSpec {
+    ) {
+      attempts = attempts + 1
+      Left("one more time")
+    }
+
+    assert(finalResult == Left("one more time"))
+    assert(attempts == 10001)
+    assert(gaveUp)
+  }
+
+  behavior of "retryingOnFailuresAndAllErrors"
+
+  it should "retry until the action succeeds" in new TestContext {
+    val policy = RetryPolicies.constantDelay[StringOr](1.second)
+
+    val finalResult = retryingOnFailuresAndAllErrors[String](
+      policy,
+      _ == "yay",
+      onError,
+      onError
+    ) {
+      attempts = attempts + 1
+      if (attempts < 3)
+        Left("one more time")
+      else
+        Right("yay")
+    }
+
+    assert(finalResult == Right("yay"))
+    assert(attempts == 3)
+    assert(errors.toList == List("one more time", "one more time"))
+    assert(!gaveUp)
+  }
+
+  /*
+   * Not sure what you were intending to test here?
+   *
+   * This is a test for `retryingOnFailuresAndAllErrors`,
+   * so there is no concept of the error being "worth retrying".
+   *
+   * This is one of the tests that was causing OOM.
+   * The action can never succeed (it always returns a Left),
+   * and the `constantDelay` retry policy will never give up,
+   * so it will keep retrying until the list of accumulated errors
+   * fills the heap.
+   */
+  //it should "retry only if the error is worth retrying" in new TestContext {
+  //  val policy = RetryPolicies.constantDelay[StringOr](1.second)
+
+  //  val finalResult = retryingOnFailuresAndAllErrors[String](
+  //    policy,
+  //    _ == "one more time",
+  //    onError,
+  //    onError
+  //  ) {
+  //    attempts = attempts + 1
+  //    if (attempts < 3)
+  //      Left("one more time")
+  //    else
+  //      Left("nope")
+  //  }
+
+  //  assert(finalResult == Left("nope"))
+  //  assert(attempts == 3)
+  //  assert(errors.toList == List("one more time", "one more time"))
+  //  assert(!gaveUp) // false because onError is only called when the error is worth retrying
+  //}
+
+  it should "retry until the policy chooses to give up" in new TestContext {
+    val policy = RetryPolicies.limitRetries[StringOr](2)
+
+    val finalResult = retryingOnFailuresAndAllErrors[String](
+      policy,
+      _ == "will never happen",
+      onError,
+      onError
+    ) {
+      attempts = attempts + 1
+      Left("one more time")
+    }
+
+    assert(finalResult == Left("one more time"))
+    assert(attempts == 3)
+    assert(
+      errors.toList == List("one more time", "one more time", "one more time")
+    )
+    assert(gaveUp)
+  }
+
+  it should "retry in a stack-safe way" in new TestContext {
+    val policy = RetryPolicies.limitRetries[StringOr](10000)
+
+    val finalResult = retryingOnFailuresAndAllErrors[String](
+      policy,
+      _ == "will never happen",
+      onError,
+      onError
+    ) {
+      attempts = attempts + 1
+      Left("one more time")
+    }
+
+    assert(finalResult == Left("one more time"))
+    assert(attempts == 10001)
+    assert(gaveUp)
+  }
+
```

Comment:
Same here, a test for `retryingOnFailuresAndAllErrors`'s handling of the failure case (not the error case) would be nice.

Reply:
[ditto](https://github.com/cb372/cats-retry/pull/259#discussion_r543570588)